File: Sources/EMKECoordinator/PCMLevelMeter.swift
import Foundation

public enum PCMLevelMeterError: Error, Equatable, Sendable {
    case oddByteCount
    case invalidSampleRate
}

public struct PCMLevelMeter: Sendable {
    public private(set) var level = 0.0

    private let noiseFloor: Double
    private let ceiling: Double
    private let attackSeconds: Double
    private let releaseSeconds: Double

    public init(
        noiseFloor: Double = 0.0015,
        ceiling: Double = 0.1,
        attackSeconds: Double = 0.08,
        releaseSeconds: Double = 0.22
    ) {
        self.noiseFloor = noiseFloor
        self.ceiling = ceiling
        self.attackSeconds = attackSeconds
        self.releaseSeconds = releaseSeconds
    }

    @discardableResult
    public mutating func observe(
        _ pcm16: Data,
        sampleRate: Double = 24_000
    ) throws -> Double {
        guard pcm16.count.isMultiple(of: 2) else {
            throw PCMLevelMeterError.oddByteCount
        }
        guard sampleRate > 0 else {
            throw PCMLevelMeterError.invalidSampleRate
        }

        let sampleCount = pcm16.count / MemoryLayout<Int16>.size
        guard sampleCount > 0 else { return level }

        let sumOfSquares = pcm16.withUnsafeBytes { bytes in
            (0..<sampleCount).reduce(into: 0.0) { sum, index in
                let sample = Int16(littleEndian: bytes.loadUnaligned(
                    fromByteOffset: index * MemoryLayout<Int16>.size,
                    as: Int16.self
                ))
                let normalized = Double(sample) / Double(Int16.max)
                sum += normalized * normalized
            }
        }
        let rms = sqrt(sumOfSquares / Double(sampleCount))
        let safeNoiseFloor = max(noiseFloor, .leastNonzeroMagnitude)
        let safeCeiling = max(ceiling, safeNoiseFloor)
        let floorDB = 20 * log10(safeNoiseFloor)
        let ceilingDB = 20 * log10(safeCeiling)
        let rmsDB = rms > 0 ? 20 * log10(rms) : -.infinity
        let rangeDB = max(ceilingDB - floorDB, .leastNonzeroMagnitude)
        let target = min(max((rmsDB - floorDB) / rangeDB, 0), 1)
        let duration = Double(sampleCount) / sampleRate
        let timeConstant = target > level ? attackSeconds : releaseSeconds
        let alpha = 1 - exp(-duration / timeConstant)
        level += (target - level) * alpha
        level = min(max(level, 0), 1)
        return level
    }

    public mutating func reset() {
        level = 0
    }
}

File: Windows/src/EMKE.Routing/PcmLevelMeter.cs
namespace EMKE.Routing;

public sealed class PcmLevelMeter
{
    private readonly double _noiseFloor;
    private readonly double _ceiling;
    private readonly double _attackSeconds;
    private readonly double _releaseSeconds;

    public PcmLevelMeter(
        double noiseFloor = 0.0015,
        double ceiling = 0.1,
        double attackSeconds = 0.08,
        double releaseSeconds = 0.22)
    {
        if (!double.IsFinite(noiseFloor) || noiseFloor < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(noiseFloor));
        }

        if (!double.IsFinite(ceiling) || ceiling < 0)
        {
            throw new ArgumentOutOfRangeException(nameof(ceiling));
        }

        if (!double.IsFinite(attackSeconds) || attackSeconds <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(attackSeconds));
        }

        if (!double.IsFinite(releaseSeconds) || releaseSeconds <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(releaseSeconds));
        }

        _noiseFloor = noiseFloor;
        _ceiling = ceiling;
        _attackSeconds = attackSeconds;
        _releaseSeconds = releaseSeconds;
    }

    public double Level { get; private set; }

    public double Observe(ReadOnlySpan<byte> pcm16, double sampleRate = 24_000)
    {
        if ((pcm16.Length & 1) != 0)
        {
            throw new ArgumentException(
                "PCM16 input must contain an even number of bytes.",
                nameof(pcm16));
        }

        if (!double.IsFinite(sampleRate) || sampleRate <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(sampleRate));
        }

        int sampleCount = pcm16.Length / sizeof(short);
        if (sampleCount == 0)
        {
            return Level;
        }

        double sumOfSquares = 0;
        for (int index = 0; index < pcm16.Length; index += 2)
        {
            ushort bits = (ushort)(pcm16[index] | (pcm16[index + 1] << 8));
            short sample = unchecked((short)bits);
            double normalized = sample / (double)short.MaxValue;
            sumOfSquares += normalized * normalized;
        }

        double rms = Math.Sqrt(sumOfSquares / sampleCount);
        double safeNoiseFloor = Math.Max(_noiseFloor, double.Epsilon);
        double safeCeiling = Math.Max(_ceiling, safeNoiseFloor);
        double floorDecibels = 20 * Math.Log10(safeNoiseFloor);
        double ceilingDecibels = 20 * Math.Log10(safeCeiling);
        double rmsDecibels = rms > 0
            ? 20 * Math.Log10(rms)
            : double.NegativeInfinity;
        double rangeDecibels = Math.Max(
            ceilingDecibels - floorDecibels,
            double.Epsilon);
        double target = Math.Clamp(
            (rmsDecibels - floorDecibels) / rangeDecibels,
            0,
            1);
        double duration = sampleCount / sampleRate;
        double timeConstant = target > Level
            ? _attackSeconds
            : _releaseSeconds;
        double alpha = 1 - Math.Exp(-duration / timeConstant);
        Level += (target - Level) * alpha;
        Level = Math.Clamp(Level, 0, 1);
        return Level;
    }

    public void Reset()
    {
        Level = 0;
    }
}
